### PR TITLE
Jakarta Security 3.0: Check for required parameters in token response

### LINKED ISSUE #24350: Access is granted when the id_token is not included in the json response from the token endpoint
I have a test token endpoint that is not including an id_token in its response.
`{"access_token":"eyJraWQiOiIiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwOi8vbG9jYWxob3N0Ojg5MzAvVG9rZW5FbmRwb2ludFNlcnZsZXQiLCJpYXQiOjE2NzY1NzY0NjMsImV4cCI6MTY3NjU4MzY2Mywic2NvcGUiOiJvcGVuaWQgcHJvZmlsZSIsInN1YiI6InRlc3R1c2VyIiwicmVhbG1OYW1lIjoiQmFzaWNSZWFsbSIsInRva2VuX3R5cGUiOiJCZWFyZXIiLCJ0b2tlbl9zcmMiOiJ0ZXN0Y2FzZSBidWlsZGVyIiwiZ3JvdXBJZHMiOlsiYWxsIl0sImF1ZCI6ImNsaWVudF8xIiwiYXRfaGFzaCI6ImR1bW15X2hhc2hfdmFsdWUifQ.XXw_ez-AX9JkzX5mxBc9A1FdKKKLxF_2jsbu6HFW0rc","token_type":"Bearer","expires_in":7199,"scope":"openid profile","refresh_token":"21MhoIC95diaQo9tb5UpFBDFlHh45NixhcKkCwRipszH6WIzKz"}`
Access to a protected app is granted.  We should fail processing the response from the token endpoint since id_token is required.

### LINKED ISSUE #24354: Access is granted when token_type is not included in the json response from the token endpoint
I have a test token endpoint that is not including token_type in its response.
`{"access_token":"eyJraWQiOiIiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwOi8vbG9jYWxob3N0Ojg5MzAvVG9rZW5FbmRwb2ludFNlcnZsZXQiLCJpYXQiOjE2NzY1ODE5MTcsImV4cCI6MTY3NjU4OTExNywic2NvcGUiOiJvcGVuaWQgcHJvZmlsZSIsInN1YiI6InRlc3R1c2VyIiwicmVhbG1OYW1lIjoiQmFzaWNSZWFsbSIsInRva2VuX3R5cGUiOiJCZWFyZXIiLCJ0b2tlbl9zcmMiOiJ0ZXN0Y2FzZSBidWlsZGVyIiwiZ3JvdXBJZHMiOlsiYWxsIl0sImF1ZCI6ImNsaWVudF8xIiwiYXRfaGFzaCI6ImR1bW15X2hhc2hfdmFsdWUifQ.ig1Dl0l7Mek179yp1qwFz53HYGCpHK24VSmMJNV_7zc","expires_in":7199,"scope":"openid profile","refresh_token":"21MhoIC95diaQo9tb5UpFBDFlHh45NixhcKkCwRipszH6WIzKz","id_token":"eyJraWQiOiIiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwOi8vbG9jYWxob3N0Ojg5MzAvVG9rZW5FbmRwb2ludFNlcnZsZXQiLCJpYXQiOjE2NzY1ODE5MTcsImV4cCI6MTY3NjU4OTExNywic2NvcGUiOiJvcGVuaWQgcHJvZmlsZSIsInN1YiI6InRlc3R1c2VyIiwicmVhbG1OYW1lIjoiQmFzaWNSZWFsbSIsInRva2VuX3R5cGUiOiJCZWFyZXIiLCJ0b2tlbl9zcmMiOiJ0ZXN0Y2FzZSBidWlsZGVyIiwiZ3JvdXBJZHMiOlsiYWxsIl0sImF1ZCI6ImNsaWVudF8xIiwiYXRfaGFzaCI6ImR1bW15X2hhc2hfdmFsdWUifQ.ig1Dl0l7Mek179yp1qwFz53HYGCpHK24VSmMJNV_7zc"}
[missingTokenType.zip](https://github.com/OpenLiberty/open-liberty/files/10760931/missingTokenType.zip)
`
Access to a protected app is granted. We should fail processing the response from the token endpoint since token_type is required.

### PR DESCRIPTION
**Pull request by ayoho**
Resolves #24350
Resolves #24354

### CODE REVIEW

**Comment by ayoho** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.security.jakartasec.3.0.internal_fat,io.openliberty.security.jakartasec.3.0.internal_fat.config,io.openliberty.security.jakartasec.3.0.internal_fat.logout,io.openliberty.security.jakartasec.3.0.internal_fat.refresh

File: dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops [hunk 1]
```diff
@@ -125,3 +122,7 @@ TOKEN_CLAIM_VALUE_MISMATCH.useraction=Ensure the attribute in the OpenID Connect
 TOKEN_MISSING_CLAIMS=CWWKS2425E: The {0} OpenID Connect client token is invalid. The jwtClaims value on the JwtContext object is null or empty.
 TOKEN_MISSING_CLAIMS.explanation=Information is missing from the token and validation of the token cannot be completed.
 TOKEN_MISSING_CLAIMS.useraction=Review the log for any earlier errors for details on the failure. Check the format and content of the token that is returned from the OpenID Connect provider.
+
+TOKEN_RESPONSE_MISSING_PARAMETER=CWWKS2426E: The token response from the OpenID Connect provider is missing at least one required parameter: {0}.
+TOKEN_RESPONSE_MISSING_PARAMETER.explanation=The OpenID Connect provider failed to return a required parameter, or the OpenID Connect provider failed to process the token request.
+TOKEN_RESPONSE_MISSING_PARAMETER.useraction=Verify that the token endpoint that is configured is correct and capable of responding to OAuth token requests.
```
Comment (ramkumar-k-9286) [inline dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops]:
```suggestion
TOKEN_RESPONSE_MISSING_PARAMETER.useraction=Verify that the configured token endpoint is correct and capable of responding to OAuth token requests.
```

File: dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops [hunk 2]
```diff
@@ -125,3 +122,7 @@ TOKEN_CLAIM_VALUE_MISMATCH.useraction=Ensure the attribute in the OpenID Connect
 TOKEN_MISSING_CLAIMS=CWWKS2425E: The {0} OpenID Connect client token is invalid. The jwtClaims value on the JwtContext object is null or empty.
 TOKEN_MISSING_CLAIMS.explanation=Information is missing from the token and validation of the token cannot be completed.
 TOKEN_MISSING_CLAIMS.useraction=Review the log for any earlier errors for details on the failure. Check the format and content of the token that is returned from the OpenID Connect provider.
+
+TOKEN_RESPONSE_MISSING_PARAMETER=CWWKS2426E: The token response from the OpenID Connect provider is missing at least one required parameter: {0}.
+TOKEN_RESPONSE_MISSING_PARAMETER.explanation=The OpenID Connect provider failed to return a required parameter, or the OpenID Connect provider failed to process the token request.
```
Comment (ramkumar-k-9286) [inline dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops]:
```suggestion
TOKEN_RESPONSE_MISSING_PARAMETER.explanation=Either the OpenID Connect provider did not return a required parameter or failed to process the token request.
```

**Comment by ayoho** [comment]
#libby

**Comment by posmith** [comment]
L2 message review completed

**Comment by ayoho** [comment]
#libby

Merged with Aruna's changes in https://github.com/OpenLiberty/open-liberty/pull/24355 to avoid a conflict in the NLS file.

**Comment by ayoho** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.security.jakartasec.3.0.internal_fat,io.openliberty.security.jakartasec.3.0.internal_fat.config,io.openliberty.security.jakartasec.3.0.internal_fat.logout,io.openliberty.security.jakartasec.3.0.internal_fat.refresh

**Comment by ayoho** [comment]
#libby

Forgot to save the unit test changes before I pushed my latest updates. Latest commit has those changes.

File: dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/exceptions/TokenResponseException.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    public TokenResponseException(String nlsMessage) {
+        this.nlsMessage = nlsMessage;
+    }
+
+    @Override
+    public String getMessage() {
+        return Tr.formatMessage(tc, "TOKEN_RESPONSE_EXCEPTION", nlsMessage);
+    }
+
+}
```
Comment (arunavemulapalli) [inline dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/exceptions/TokenResponseException.java]: are we using this exception class?

**Review by arunavemulapalli — DISMISSED** [review]

File: dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/exceptions/TokenResponseException.java (hunk 3 above)
Comment (ayoho) [reply dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/exceptions/TokenResponseException.java]: Turns out no, we are not. I had planned to use this for exceptions with the token response, but I just ended up using the existing classes. Good catch. I'll take this out in another PR.

**Comment by ayoho** [comment]
#libby
#build
#spawn.fullfat.buckets=io.openliberty.security.jakartasec.3.0.internal_fat,io.openliberty.security.jakartasec.3.0.internal_fat.config,io.openliberty.security.jakartasec.3.0.internal_fat.logout,io.openliberty.security.jakartasec.3.0.internal_fat.refresh

**Review by arunavemulapalli — APPROVED** [review]

**Review by c00crane — APPROVED** [review]

**Pull request merged**